src/compiler/moduleNameResolver.ts

```diff
@@ -2413,8 +2413,8 @@ namespace ts {
                 );
             if (
                 !pathAndExtension && packageInfo
-                && packageInfo.packageJsonContent.exports === undefined
-                && packageInfo.packageJsonContent.main === undefined
+                // eslint-disable-next-line no-null/no-null
+                && (packageInfo.packageJsonContent.exports === undefined || packageInfo.packageJsonContent.exports === null)

```

Comment:
Double checking on https://nodejs.org/api/esm.html#resolution-algorithm, recognizing both null and undefined here for `exports` seems good (only one usage makes that explicit, but it should be fine), however ignoring `main` and looking up an index regardless is not what `PACKAGE_RESOLVE` describes for an `export`less package... it just says `Return the URL resolution of packageSubpath in packageURL.`, which doesn't imply any fallback to `index.js` lookup if the `main` fails. If in practice it is, the algorithm in the `node` docs may be wrong. (Which doesn't surprise me, if I'm honest, I think the people writing the code are more pragmatic than the people reviewing/writing the docs.)

Reply:
Indeed, there's a large [legacy main resolution](https://github.com/nodejs/node/blob/d859e9e997a3de9bdd7e116b2878acc2a3cf4bd6/lib/internal/modules/esm/resolve.js#L280) code block which isn't included in the documented algorithm.

Reply:
I've opened https://github.com/nodejs/node/issues/43269 for the docs issue on node.